src/frontegg-auth/src/auth.rs

```diff
@@ -314,206 +230,250 @@ impl Authentication {
-                    break;
+impl AuthenticatorInner {
+    async fn authenticate(
+        self: &Arc<Self>,
+        expected_email: String,
+        password: AppPassword,
+    ) -> Result<AuthSessionHandle, Error> {
+        // Attempt initial app password exchange.
+        let mut claims = self
+            .exchange_app_password(&expected_email, password)
+            .await?;
+
+        // Prep session information.
+        let ident = Arc::new(AuthSessionIdent {
+            user: claims.email.clone(),
+            tenant_id: claims.tenant_id,
+        });
+        let external_metadata = claims.to_external_user_metadata();
+        let (external_metadata_tx, external_metadata_rx) = watch::channel(external_metadata);
+        let external_metadata_tx = Arc::new(external_metadata_tx);
+
+        // Store session to make it available for future requests to latch on
+        // to.
+        {
+            let mut sessions = self.sessions.lock().expect("lock poisoned");
+            sessions.insert(
+                password,
+                AuthSession::Active {
+                    ident: Arc::clone(&ident),
+                    external_metadata_tx: Arc::clone(&external_metadata_tx),
                 },
-            }
+            );
+        }
+
+        // Start background refresh task.
+        let name = format!("frontegg-auth-refresh-{}", password.client_id);
+        mz_ore::task::spawn(|| name, {
+            let inner = Arc::clone(self);
+            async move {
+                let gauge = inner.metrics.refresh_tasks_active.with_label_values(&[]);
+                gauge.inc();
+
+                loop {
+                    // Scale the validity duration by 0.8. The Frontegg Python
+                    // SDK scales the expires_in this way.
+                    //
+                    // <https://github.com/frontegg/python-sdk/blob/840f8318aced35cea6a41d83270597edfceb4019/frontegg/common/frontegg_authenticator.py#L45>
+                    let valid_for = claims.exp - inner.now.as_secs();
+                    let valid_for = f64::cast_lossy(valid_for) * 0.8;
+                    let valid_for = u64::cast_lossy(valid_for);
+
+                    if valid_for < 60 {
+                        warn!(%valid_for, "unexpectedly low token validity");
+                    }
 
-            let current_response = client
-                .refresh_token(&refresh_url, &latest_response.refresh_token, &metrics)
-                .await;
+                    trace!(%valid_for, "waiting for token validity period");
 
-            // If we failed to refresh, then close the channel.
-            let current_response = match current_response {
-                Ok(resp) => {
-                    tracing::debug!(?id, "refresh successful");
-                    resp
+                    // Wait out validity duration.
+                    time::sleep(Duration::from_secs(valid_for)).await;
```

Comment:
This will leak a lotta resources in balancer if sessions are not shared.

Reply:
(They should be shared!)